### ISSUE TITLE
Feature Request: Progress info back to main thread

### ISSUE THREAD

**Issue by KrishnaPG** [andywer/threads.js#5]
Long running workers should be able to inform their progress to the main thread. For example,
````
spawn(...)
.on('progress', function(current, total) { 
   console.log(current/total + " % complete");
});
````

If this is already achievable in the current implementation, a simple example in Readme would be of great help.

**Comment by andywer** [comment]
Yeah, that's a good idea. It's a bit tricky, because threads.js does not know about the semantics of the worker's job. But I think it's ok if you are just able to trigger a progress callback manually from your worker code. Your main thread code would look like this:

```js
.on('progress', function(progressValueYourWorkerSent) {
  ...
});
```

I think that should do the trick. Tell me what you think about it and I try to implement it as soon as I have some time.

**Comment by KrishnaPG** [comment]
Yes. That should do. The worker can manually raise the progress event (for whoever is listening) and keep doing its business. (question: would it be possible to raise that event asynchronously by the worker by default, and also synchronously if chosen to be so??)

I believe when you say `function(progressValueYourWorkerSent)` the `progressValueYourWorkerSent` can be any custom user defined object (and not just % numeric value) - right? That way, user will have more control. For example, instead of accumulating the results till the end/completion, worker can keep dispatching the intermediate results in batches to the main thread (like in streams). Say, converting large array of json string to json objects - the worker can keep dispatching converted objects in batches of, say 100. 

You have a very good ["streaming" example](https://github.com/andywer/threads.js#streaming) in the readme. In that one the worker returns results only once at the end (I think). But with progress events, in combination with your streaming capability, one can easily create "CEP" engines with 'window' operations.
For example, from a continuous incoming stream of numbers/objects from main thread,
*   for every 5 seconds, keep reporting the average of numbers seen in the last 5 seconds window
*   whenever there is a sequence of 3 numbers whose sum is > 10, report those numbers
*   continuously report the mean/variance/min/max of the numbers at 1 hour intervals so on....

Libraries (such as [this](https://github.com/darach/eep-js)) can heavily benefit from such work.


On a different note on the sync/async side:

For simple progress reports of %values, the asynchronous (one-way / unacknowledged) events from the worker to the main thread should be fine. But for complex tasks, say processing a large file in the background, controlling the worker thread from main thread (such as pause/resume/getNext...) would be made possible by synchronous (two-way / acknowledged) events between worker and threads. 

Would that kind of sync/async events between web-wroker and main thread possible in the current standard? What about the node.js?

**Comment by andywer** [comment]
Since everything is running asynchronously, there cannot be synchronous progress handling.

Yeah, you could probably pass pretty much everything to the progress callback, but you can simply stick to the "streaming" example you pointed out. The thread can actually respond multiple times. So in the example it will return once for every `.send()`.

But I think there is no way to pause/resume a thread (at least for web workers). You can only send some data to your thread, stating that it should pause and you would have to do pausing your current actions manually in the thread code.

So currently there is only `.send()` on the main thread's side and calling the callback (perhaps multiple times) on the child thread's side. Node.js as well as in browsers.

**Comment by KrishnaPG** [comment]
Thanks @andywer . The manual pause/resume in the worker should do fine.

As for the sync/async - I meant the execution of the callback from the child-thread. For example, when the worker thread makes a progress report to main thread (using callback?), does the worker thread "halt" (as in wait for the callback to complete and return) or does it just issue the callback (asynchronously somehow) and keep doing further business?

**Comment by andywer** [comment]
Alright then. I am going to implement some progress update thingy.

And no, there is no halt. It is really a multi-threading library ;) So it's about code running in parallel on different CPU cores and passing messages between them.

I will kerp you updated!

**Comment by andywer** [comment]
What do you think? You will find a sample code snippet in the README file (develop branch).

https://github.com/andywer/threads.js/commit/3c3890ae7bd0a31450d0e7ce3045aada7509f1af

**Comment by KrishnaPG** [comment]
Thanks much for the quick response. The implementations looks good on the first glance. Will let you know in case of any problems. But looks like you nailed it and no problems may come. Closing this for now.

**Comment by andywer** [comment]
Just released version 0.5.3 containing the feature.
